File: src/granola/formatters/combined.py
"""Combined notes and transcript formatting."""

from datetime import datetime

from granola.cache.reader import TranscriptSegment


def format_combined(
    title: str,
    doc_id: str,
    created_at: str,
    updated_at: str,
    notes_content: str,
    segments: list[TranscriptSegment],
    folders: list[str],
) -> str:
    """Format notes and transcript into a single text file.

    Args:
        title: Document title.
        doc_id: Document ID.
        created_at: Creation timestamp.
        updated_at: Update timestamp.
        notes_content: Plain text notes content.
        segments: Transcript segments.
        folders: List of folder names.

    Returns:
        Combined formatted string.
    """
    lines: list[str] = []

    # Header
    lines.append("=" * 80)

    if title:
        lines.append(title)

    lines.append(f"ID: {doc_id}")

    if created_at:
        lines.append(f"Created: {created_at}")

    if updated_at:
        lines.append(f"Updated: {updated_at}")

    if folders:
        lines.append(f"Folders: {', '.join(folders)}")

    lines.append("=" * 80)

    # Notes section
    lines.append("")
    lines.append("## Notes")
    lines.append("")

    if notes_content and notes_content.strip():
        lines.append(notes_content)
    else:
        lines.append("(No notes)")

    # Transcript section
    lines.append("")
    lines.append("=" * 80)
    lines.append("")
    lines.append("## Transcript")
    lines.append("")

    if segments:
        for segment in segments:
            timestamp = _parse_timestamp(segment.start_timestamp)
            speaker = "You" if segment.source == "microphone" else "System"
            lines.append(f"[{timestamp}] {speaker}: {segment.text}")
    else:
        lines.append("(No transcript available)")

    return "\n".join(lines)


def _parse_timestamp(timestamp: str) -> str:
    """Convert ISO 8601 timestamp to HH:MM:SS format.

    Args:
        timestamp: ISO 8601 timestamp string.

    Returns:
        Formatted time string or original on error.
    """
    try:
        ts = timestamp.replace("Z", "+00:00")
        dt = datetime.fromisoformat(ts)
        return dt.strftime("%H:%M:%S")
    except ValueError:
        return timestamp

File: internal/formatter/combined.go
// Package formatter provides functionality for formatting combined notes and transcripts.
package formatter

import (
	"fmt"
	"strings"
	"time"

	"github.com/theantichris/granola/internal/cache"
)

// FormatCombined formats notes and transcript into a single text file.
func FormatCombined(title, id, createdAt, updatedAt, notesContent string, segments []cache.TranscriptSegment, folders []string) string {
	var builder strings.Builder

	// Header
	builder.WriteString(strings.Repeat("=", 80))
	builder.WriteString("\n")

	if title != "" {
		builder.WriteString(title)
		builder.WriteString("\n")
	}

	builder.WriteString("ID: ")
	builder.WriteString(id)
	builder.WriteString("\n")

	if createdAt != "" {
		builder.WriteString("Created: ")
		builder.WriteString(createdAt)
		builder.WriteString("\n")
	}

	if updatedAt != "" {
		builder.WriteString("Updated: ")
		builder.WriteString(updatedAt)
		builder.WriteString("\n")
	}

	if len(folders) > 0 {
		builder.WriteString("Folders: ")
		builder.WriteString(strings.Join(folders, ", "))
		builder.WriteString("\n")
	}

	builder.WriteString(strings.Repeat("=", 80))
	builder.WriteString("\n")

	// Notes section
	builder.WriteString("\n## Notes\n\n")
	if strings.TrimSpace(notesContent) != "" {
		builder.WriteString(notesContent)
		builder.WriteString("\n")
	} else {
		builder.WriteString("(No notes)\n")
	}

	// Transcript section
	builder.WriteString("\n")
	builder.WriteString(strings.Repeat("=", 80))
	builder.WriteString("\n")
	builder.WriteString("\n## Transcript\n\n")

	if len(segments) > 0 {
		for _, segment := range segments {
			startTime := parseTimestamp(segment.StartTimestamp)
			speaker := "System"
			if segment.Source == "microphone" {
				speaker = "You"
			}
			builder.WriteString(fmt.Sprintf("[%s] %s: %s\n", startTime, speaker, segment.Text))
		}
	} else {
		builder.WriteString("(No transcript available)\n")
	}

	return builder.String()
}

// parseTimestamp converts ISO 8601 timestamp to HH:MM:SS format.
func parseTimestamp(timestamp string) string {
	t, err := time.Parse(time.RFC3339, timestamp)
	if err != nil {
		return timestamp
	}
	return t.Format("15:04:05")
}
